File: java-app/src/main/java/com/showcase/java/slice/genericsrecords/TypeSafeIds.java
package com.showcase.java.slice.genericsrecords;

import com.showcase.java.domain.model.ProjectId;
import com.showcase.java.domain.model.TaskId;
import com.showcase.java.domain.model.UserId;

import java.util.List;

/**
 * Slice 9 — Generics & Records: TypeSafeIds
 *
 * Java equivalent of Kotlin value classes for type-safe identifiers.
 * Java records provide similar wrapper semantics and type safety,
 * but with runtime allocation overhead (records are heap-allocated objects).
 *
 * Kotlin value classes are erased at runtime — TaskId("task-1") becomes
 * just the String "task-1" on the JVM. Java records always allocate.
 */
public final class TypeSafeIds {

    private TypeSafeIds() {}

    /**
     * Demonstrates that records prevent mixing up identifier types.
     * Passing a UserId where a TaskId is expected is a compile-time error.
     *
     * // This would NOT compile:
     * // TaskId taskId = new UserId("user-1");  // Type mismatch!
     */
    public static String lookupTask(TaskId taskId) {
        return "Looking up task: " + taskId.value();
    }

    public static String lookupUser(UserId userId) {
        return "Looking up user: " + userId.value();
    }

    public static String lookupProject(ProjectId projectId) {
        return "Looking up project: " + projectId.value();
    }

    /**
     * Demonstrates that records work in collections with type safety.
     * Unlike Kotlin value classes, each record instance is heap-allocated.
     */
    public static List<TaskId> filterTaskIds(List<TaskId> ids, String prefix) {
        return ids.stream()
                .filter(id -> id.value().startsWith(prefix))
                .toList();
    }

    /**
     * Record for returning a triple of identifiers.
     * Kotlin equivalent uses Triple&lt;TaskId, ProjectId, UserId&gt;.
     */
    public record IdTriple(TaskId taskId, ProjectId projectId, UserId userId) {}

    /**
     * Demonstrates creating record-based identifiers.
     * Unlike Kotlin value classes, these are heap-allocated at runtime.
     */
    public static IdTriple createIds() {
        return new IdTriple(
                new TaskId("task-001"),
                new ProjectId("proj-001"),
                new UserId("user-001")
        );
    }
}

File: kotlin-app/src/main/kotlin/com/showcase/kotlin/slice/inlinereifiedvalue/TypeSafeIds.kt
package com.showcase.kotlin.slice.inlinereifiedvalue

import com.showcase.kotlin.domain.model.ProjectId
import com.showcase.kotlin.domain.model.TaskId
import com.showcase.kotlin.domain.model.UserId

/**
 * Slice 9 — Value Classes: TypeSafeIds
 *
 * Demonstrates Kotlin value classes for type-safe identifiers with
 * zero runtime overhead. The value classes are defined in the domain
 * model (TaskId, ProjectId, UserId) — this file shows usage patterns
 * that highlight the compile-time safety they provide.
 *
 * Java records provide similar wrapper semantics but with runtime
 * allocation overhead. Kotlin value classes are erased at runtime.
 */
object TypeSafeIds {

    /**
     * Demonstrates that value classes prevent mixing up identifier types.
     * Passing a UserId where a TaskId is expected is a compile-time error.
     *
     * // This would NOT compile:
     * // val taskId: TaskId = UserId("user-1")  // Type mismatch!
     */
    fun lookupTask(taskId: TaskId): String =
        "Looking up task: ${taskId.value}"

    fun lookupUser(userId: UserId): String =
        "Looking up user: ${userId.value}"

    fun lookupProject(projectId: ProjectId): String =
        "Looking up project: ${projectId.value}"

    /**
     * Demonstrates that value classes work seamlessly in collections
     * and higher-order functions while maintaining type safety.
     */
    fun filterTaskIds(ids: List<TaskId>, prefix: String): List<TaskId> =
        ids.filter { it.value.startsWith(prefix) }

    /**
     * Demonstrates creating value class instances — zero overhead at runtime.
     * At the JVM level, TaskId("task-1") is just the String "task-1".
     */
    fun createIds(): Triple<TaskId, ProjectId, UserId> = Triple(
        TaskId("task-001"),
        ProjectId("proj-001"),
        UserId("user-001")
    )
}
